cmd/xl-packs.go

```diff
@@ -0,0 +1,437 @@
+ */
+
+package cmd
+
+import (
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/hash"
+	"golang.org/x/sync/errgroup"
+)
+
+// xlPacks implements Objectlayer combining a static list of erasure coded
+// object layers. NOTE: There is no dynamic scaling allowed or intended in
+// current design.
+type xlPacks struct {
+	layers []*xlObjects
+
+	// Pack level listObjects pool management.
+	listPool *treeWalkPool
+}
+
+// Initialize new set of erasure coded packs.
+func newXLPacks(layers []*xlObjects) ObjectLayer {
+	return &xlPacks{
+		layers:   layers,
+		listPool: newTreeWalkPool(globalLookupTimeout),
+	}
+}
+
+// StorageInfo - combines output of StorageInfo across all erasure
+// coded object layers.
+func (s xlPacks) StorageInfo() StorageInfo {
+	var info StorageInfo
+	info.Backend.Type = Erasure
+	for _, layer := range s.layers {
+		linfo := layer.StorageInfo()
+		info.Total = info.Total + linfo.Total
+		info.Free = info.Free + linfo.Free
+		info.Backend.OnlineDisks = info.Backend.OnlineDisks + linfo.Backend.OnlineDisks
+		info.Backend.OfflineDisks = info.Backend.OfflineDisks + linfo.Backend.OfflineDisks
+	}
+	return info
+}
+
+func (s xlPacks) Shutdown() error {
+	var g errgroup.Group
+
+	for _, layer := range s.layers {
+		layer := layer // https://golang.org/doc/faq#closures_and_goroutines
+		g.Go(func() error {
+			return layer.Shutdown()
+		})
+	}
+
+	return g.Wait()
+}
+
+func (s xlPacks) MakeBucketWithLocation(bucket, location string) error {
+	var g errgroup.Group
+
+	for _, layer := range s.layers {
+		layer := layer // https://golang.org/doc/faq#closures_and_goroutines
+		g.Go(func() error {
+			return layer.MakeBucketWithLocation(bucket, location)
+		})
+	}
+
+	// Wait for all makebucket to finish.
+	return g.Wait()
+}
+
+func (s xlPacks) getLoadBalancedLayers(input string) (layers []*xlObjects) {
+	// Based on the random shuffling return back randomized disks.
+	for _, i := range hashOrder(input, len(s.layers)) {
+		layers = append(layers, s.layers[i-1])
+	}
+	return layers
+}
+
+func (s xlPacks) getHashedLayer(input string) (layers *xlObjects) {
+	return s.getLoadBalancedLayers(input)[0]
+}
+
+func (s xlPacks) GetBucketInfo(bucket string) (bucketInfo BucketInfo, err error) {
+	for _, layer := range s.getLoadBalancedLayers(UTCNow().String()) {
+		bucketInfo, err = layer.GetBucketInfo(bucket)
+		if err != nil {
+			return bucketInfo, err
+		}
+		break
+	}
+	return bucketInfo, nil
+}
+
+func (s xlPacks) DeleteBucket(bucket string) error {
+	var g errgroup.Group
+
+	for _, layer := range s.layers {
+		layer := layer // https://golang.org/doc/faq#closures_and_goroutines
+		g.Go(func() error {
+			return layer.DeleteBucket(bucket)
+		})
+	}
+
+	return g.Wait()
+}
+
+func (s xlPacks) ListBuckets() (buckets []BucketInfo, err error) {
+	for _, layer := range s.getLoadBalancedLayers(UTCNow().String()) {
+		buckets, err = layer.ListBuckets()
+		if err != nil {
+			return nil, err
+		}
+		break
```

Comment:
for loop breaks during first iteration itself. I think you wanted to move to the next pack till we get a good result.

Reply:
we already got a good result the first good result @krishnasrinivas

Reply:
OK, if there is no purpose for the `for-loop` then:

```
func (s xlPacks) ListBuckets() (buckets []BucketInfo, err error) {
        for _, layer := range s.getLoadBalancedLayers(UTCNow().String()) {
                buckets, err = layer.ListBuckets()
                if err != nil {
                        return nil, err
                }
                break
        }
        return buckets, nil
}
```

can be replaced with:
```
func (s xlPacks) ListBuckets() (buckets []BucketInfo, err error) {
        return s.getLoadBalancedLayers(UTCNow().String())[0]
}
```